File: 039. 组合总和.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.res = []

        def backtrack(path, start_index, curr_sum):
            if curr_sum > target:
                return
            if curr_sum == target:
                self.res.append(path[:])
                return

            for i in range(start_index, len(candidates)):
                num = candidates[i]
                path.append(num)
                backtrack(path, i, curr_sum + num)
                path.pop()

        backtrack([], 0, 0)
        return self.res

File: 039. 组合总和.go
package main

func combinationSum(candidates []int, target int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)

	var backtrack func(startIdx, currSum int)
	backtrack = func(startIdx, currSum int) {
		if currSum > target {
			return
		}
		if currSum == target {
			temp := make([]int, len(path))
			copy(temp, path)
			res = append(res, temp)
		}

		for i := startIdx; i < len(candidates); i++ {
			path = append(path, candidates[i])
			backtrack(i, currSum+candidates[i])
			path = path[:len(path)-1]
		}
	}

	backtrack(0, 0)
	return res
}
